### PR TITLE
🐳 Docker support

### PR DESCRIPTION
**Pull request by nchristopher**
This pull request adds **Docker support** to Needle, allowing easy deploys of the bot.

Automatic building & deploying for __all__ releases and commits - along with just building for pull requests - is done using GitHub actions. In order to prevent building the same thing twice, and to skip pushes if builds fail, this new action replaces the previous build action. Images are pushed to GitHub's container registry.

Additionally, an example `docker-compose.yml` was created for Docker Compose users.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+COPY . .
+
+RUN npm run build
+
+ENV CONFIGS_PATH=/configs
+
+VOLUME [ "/configs" ]
+
+CMD ["node", "./dist/index.js"]
```
Comment (c43721) [inline Dockerfile:17]: User should be `node` in Dockerfiles to not be in root

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,17 @@
+FROM node:16
+
+WORKDIR /app
+
+COPY package*.json ./
+
+RUN npm ci
+
+COPY . .
```
Comment (c43721) [inline Dockerfile:9]: You're copying everything here. Do a multi-stage build (1 for the TS and the other for running JUST compiled JS)

File: .github/workflows/docker.yml [hunk 3]
```diff
@@ -0,0 +1,24 @@
+name: Build and publish a Docker image to ghcr.io
```
Comment (c43721) [inline .github/workflows/docker.yml:1]: Can we prevent this from running if the build fails?

**Review by c43721 — CHANGES_REQUESTED** [review]
Docker nits + question

See [my dockerfile](https://github.com/payload-bot/payload-neo/blob/main/Dockerfile) for a bit more context

**Comment by c43721** [comment]
FYI the docker changes are significant - you can really bloat the image by having unnecessary files which also increases build time.

I'd like to test this as well, there's docker layer caching which can significantly improve or ruin build times if your steps are out of order.

File: .github/workflows/docker.yml (hunk 3 above)
Comment (nchristopher) [reply .github/workflows/docker.yml:1]: If you mean prevent running if the normal TypeScript action build fails, I don't think that's possible. We'd have to merge the two GitHub actions together.

Comment (c43721) [reply .github/workflows/docker.yml:1]: I think we can, I'll do some poking

File: Dockerfile [hunk 4]
```diff
@@ -1,16 +1,23 @@
-FROM node:16
-
+# Build
+FROM node:16.14.0-slim AS build
 WORKDIR /app
 
 COPY package*.json ./
-
-RUN npm ci
+RUN npm install --production
```
Comment (c43721) [inline Dockerfile:6]: Remove this production from the build step, otherwise, your npm build will fail (since it wouldn't have types!!)

**Review by c43721 — CHANGES_REQUESTED** [review]

**Comment by nchristopher** [comment]
I'm going to look into `dev-[commit hash]` tags and a few other GitHub action related things before this can be merged; converting to a draft.

File: .github/workflows/docker.yml (hunk 3 above)
Comment (nchristopher) [reply .github/workflows/docker.yml:1]: Looking into this more, I think we can put both of those steps in one action & skip the Docker step if the build step fails. We also won't have to build each commit twice.

**Comment by c43721** [comment]
Sounds good. Good idea to take (steal) from djs as well here- I think converting their package build system to a image build system should be straight forward

File: .dockerignore [hunk 5]
```diff
@@ -0,0 +1,14 @@
+.git/
+node_modules/
+dist/
+.github/
+configs/
+branding/
+scripts/
```
Comment (c43721) [inline .dockerignore:7]: Does this register with scripts ignored? I haven't tried this yet.

File: Dockerfile [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+COPY package*.json ./
+RUN npm install --production
+COPY --from=build /app/dist ./dist
+
+VOLUME [ "/configs" ]
+
+USER node
+
+CMD ["node", "./dist/index.js"]
```
Comment (c43721) [inline Dockerfile:26]: More of a question, but @MarcusOtter do we emit source maps currently (.map.ts I think)? If so, maybe we should consider running with source maps enabled so if we get an exception, it'll give us the error from the source TS, and not the dist JS.

**Review by c43721 — APPROVED** [review]
Pretty sure this is good now. 

Docker layers look good, build.yml looks fine (though I think we lose our PR build status?)

File: Dockerfile (hunk 6 above)
Comment (MarcusOtter) [reply Dockerfile:26]: I don't think we do, I always get errors pointing at the js files, have no clue about how to get source maps working

File: .dockerignore (hunk 5 above)
Comment (nchristopher) [reply .dockerignore:7]: Commands aren't automatically deployed to Discord anyway, so ignoring scripts doesn't affect anything.

I could add a command that runs on startup though.

**Comment by nchristopher** [comment]
We don't - Docker will build on PRs (which compiles TypeScript), it just won't push PR images to the registry.

File: .dockerignore (hunk 5 above)
Comment (c43721) [reply .dockerignore:7]: I'm thinking about this- perhaps leave it out. If we change commands, we should manually update IMO.

File: Dockerfile (hunk 6 above)
Comment (c43721) [reply Dockerfile:26]: @nchristopher Let's enable sourcemaps then:

`CMD ["node", "--enable-source-maps", "."`

**Review by c43721 — CHANGES_REQUESTED** [review]
Let's enable sourcemaps, pending #54 I think

**Review by c43721 — APPROVED** [review]

**Comment by MarcusOtter** [comment]
Do you have any pointers as to how I can test this out before merging? @nchristopher @c43721 
Also, we should update the README with some docs around this

**Comment by c43721** [comment]
test by deploying it lol

not sure what you mean "testing" this? I can optimize the cache but might as well merge it into main first

**Review by MarcusOtter — COMMENTED** [review]
I think it looks good, can't say I know much about it, but the GitHub Actions build file looks good at least 😄

**Pull request merged**